File: v1/src/main/java/com/api/v1/services/book/BookRegistrationServiceImpl.java
package com.api.v1.services.book;

import com.api.v1.builders.book.BookBuilder;
import com.api.v1.domain.entities.Book;
import com.api.v1.domain.repositories.BookRepository;
import com.api.v1.dtos.requests.BookRegistrationRequestDto;
import com.api.v1.dtos.responses.BookResponseDto;
import com.api.v1.exceptions.book.DuplicatedIsbnException;
import com.api.v1.mappers.book.BookResponseMapper;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class BookRegistrationServiceImpl implements BookRegistrationService {

    @Autowired
    private BookRepository repository;

    @Override
    public Mono<BookResponseDto> register(@Valid BookRegistrationRequestDto request) {
        return repository
                .findAll()
                .filter(book -> book.getIsbn().equals(request.isbn()))
                .hasElements()
                .flatMap(exists -> {
                    if (exists) return Mono.error(new DuplicatedIsbnException(request.isbn()));
                    return Mono.defer(() -> {
                        Book newBook = BookBuilder.create().fromDto(request).build();
                        Mono<Book> savedNewBook = repository.save(newBook);
                        return savedNewBook.flatMap(savedBook -> Mono.just(BookResponseMapper.map(savedBook)));
                    });
                });
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/book/services/BookRegistrationServiceImpl.kt
package com.api.v2.book.services

import com.api.v2.book.domain.Book
import com.api.v2.book.domain.BookRepository
import com.api.v2.book.dtos.BookRegistrationRequestDto
import com.api.v2.book.dtos.BookResponseDto
import com.api.v2.book.exceptions.DuplicatedIsbnException
import com.api.v2.book.utils.BookResponseMapper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.count
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class BookRegistrationServiceImpl: BookRegistrationService {

    @Autowired
    lateinit var bookRepository: BookRepository

    override suspend fun register(requestDto: BookRegistrationRequestDto): BookResponseDto {
        return withContext(Dispatchers.IO) {
            if (bookRepository.findAll().filter { e -> e.isbn == requestDto.isbn }.count() != 0) {
                throw DuplicatedIsbnException(requestDto.isbn)
            }
            val book = Book(requestDto)
            val savedBook = bookRepository.save(book)
            BookResponseMapper.map(savedBook)
        }
    }

}